File: src/main/java/com/hsmchurch/app/noticeboard/ui/controllers/MeNoticeBoardController.java
package com.hsmchurch.app.noticeboard.ui.controllers;

import com.hsmchurch.app.noticeboard.application.NoticeService;
import com.hsmchurch.app.noticeboard.domain.Notice;
import com.hsmchurch.app.noticeboard.ui.request.NoticeUpdateRequest;
import com.hsmchurch.app.noticeboard.ui.request.NoticeUploadRequest;
import lombok.RequiredArgsConstructor;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import org.springframework.web.util.UriComponents;
import org.springframework.web.util.UriComponentsBuilder;

import static com.hsmchurch.app.security.AuthenticationHelper.currentUserId;
import static com.hsmchurch.app.security.AuthenticationHelper.currentUserIsAdmin;

@RequestMapping(path = "/api/v1/me/notices")
@RestController
@RequiredArgsConstructor
public class MeNoticeBoardController {

    private final NoticeService noticeService;

    @PostMapping
    public ResponseEntity<Void> register(@RequestBody final NoticeUploadRequest noticeUploadRequest,
                                         UriComponentsBuilder b) {
        final Notice savedNotice = noticeService.upload(noticeUploadRequest, currentUserId());
        UriComponents uriComponents = b.path("/api/v1/notices/{id}").buildAndExpand(savedNotice.getId());

        return ResponseEntity.created(uriComponents.toUri()).build();
    }

    @PutMapping("/{id}")
    public ResponseEntity<Void> update(@PathVariable final Long id,
                                       @RequestBody final NoticeUpdateRequest noticeUpdateRequest) {
        noticeService.updateNotice(noticeUpdateRequest, id, currentUserId());

        return ResponseEntity.noContent().build();
    }

    @DeleteMapping("/{id}")
    public ResponseEntity<Void> delete(@PathVariable final Long id) {
        if (currentUserIsAdmin()) {
            if (noticeService.delete(id)) return ResponseEntity.noContent().build();
            return ResponseEntity.badRequest().build();
        }
        if (noticeService.delete(id, currentUserId())) return ResponseEntity.noContent().build();
        return ResponseEntity.badRequest().build();
    }

}

File: src/main/java/com/hsmchurch/app/noticeboard/ui/controllers/MeNoticeBoardController.kt
package com.hsmchurch.app.noticeboard.ui.controllers

import com.hsmchurch.app.noticeboard.application.NoticeService
import com.hsmchurch.app.noticeboard.ui.request.NoticeUpdateRequest
import com.hsmchurch.app.noticeboard.ui.request.NoticeUploadRequest
import com.hsmchurch.app.security.AuthenticationHelper.currentUserId
import com.hsmchurch.app.security.AuthenticationHelper.currentUserIsAdmin
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*
import org.springframework.web.util.UriComponentsBuilder

@RequestMapping(path = ["/api/v1/me/notices"])
@RestController
class MeNoticeBoardController constructor(
        private val noticeService: NoticeService
) {

    @PostMapping
    fun register(@RequestBody noticeUploadRequest: NoticeUploadRequest,
                 b: UriComponentsBuilder): ResponseEntity<Void> {
        val savedNotice = noticeService.upload(noticeUploadRequest, currentUserId())
        val uriComponents = b.path("/api/v1/notices/{id}").buildAndExpand(savedNotice.id)

        return ResponseEntity.created(uriComponents.toUri()).build()
    }

    @PutMapping("/{id}")
    fun update(@PathVariable id: Long,
               @RequestBody noticeUpdateRequest: NoticeUpdateRequest): ResponseEntity<Void> {
        noticeService.updateNotice(noticeUpdateRequest, id, currentUserId())

        return ResponseEntity.noContent().build()
    }

    @DeleteMapping("/{id}")
    fun delete(@PathVariable id: Long): ResponseEntity<Void> {
        if (currentUserIsAdmin()) {
            return if (noticeService.delete(id)) ResponseEntity.noContent().build()
            else ResponseEntity.badRequest().build()
        }
        return if (noticeService.delete(id, currentUserId())) ResponseEntity.noContent().build()
        else ResponseEntity.badRequest().build()
    }

}
